api/src/main/java/edu/cornell/mannlib/vitro/webapp/controller/api/SparqlUpdateApiController.java

```diff
@@ -76,26 +82,47 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp)
 
 	private UpdateRequest parseUpdateString(HttpServletRequest req)
 			throws ParseException {
-		String update = req.getParameter("update");
-		if (StringUtils.isBlank(update)) {
-			log.debug("No update parameter.");
-			throw new ParseException("No 'update' parameter.");
+		InputStream update;
+		String reqtype = req.getContentType();
+		UsingList usingList = processProtocol(req);
+
+		// Support update via both POST body and an 'update' parameter 
+		if (reqtype.equalsIgnoreCase("application/sparql-update")) {
+			try {
+				update = req.getInputStream();
+			} catch (IOException e) {
+				log.debug("Error parsing POST body.");
+				throw new ParseException("Error parsing POST body.");
+			}
 		}
-
-		if (!StringUtils.containsIgnoreCase(update, "GRAPH") && !StringUtils.containsIgnoreCase(update, "WITH")) {
-			if (log.isDebugEnabled()) {
-				log.debug("No GRAPH or WITH uri in '" + update + "'");
+		else {
+			String updateString = req.getParameter("update");
+			if (StringUtils.isBlank(updateString)) {
+				log.debug("No update parameter.");
+				throw new ParseException("No 'update' parameter.");
+			}
+			if (!StringUtils.containsIgnoreCase(updateString, "GRAPH") && !StringUtils.containsIgnoreCase(updateString, "WITH")) {
+				if (log.isDebugEnabled()) {
+					log.debug("No GRAPH or WITH uri in '" + updateString + "'");
+				}
+				throw new ParseException("SPARQL update must specify a GRAPH ( or WITH) URI.");
+			}
+			try {
+				update = org.apache.commons.io.IOUtils.toInputStream(updateString, "UTF-8");
+			} catch (IOException e) {
+				log.debug("Error parsing POST body.");
+				throw new ParseException("Error parsing POST body.");
 			}
-			throw new ParseException("SPARQL update must specify a GRAPH ( or WITH) URI.");
 		}
 
 		try {
-			return UpdateFactory.create(update);
+			return UpdateFactory.read(usingList, update);
```

Comment:
I suggest moving the line 87 just before this line 
UsingList usingList = processProtocol(req);
There is a couple of points where an exception may be thrown, and in that case processing protocol is not necessary.